### ISSUE TITLE
Make promises part of documentation more clear

### ISSUE THREAD

**Issue by elhigu** [knex/knex#5886]
Currently documentation basically says that calling `.then` makes query builder to a promise, which is not true.

http://knexjs.org/#Interfaces-Promises

Some discussion about it is found in https://github.com/knex/knex/issues/3515#issuecomment-550201065
but the whole section should be revised.

**Comment by mercmobily** [comment]
Just adding a writeup to simplify writing of documentation...

**Current Behavior**: Knex query builders implement a "lazy evaluation" pattern - queries don't execute until you explicitly trigger them with `.then()`, `await`, or other promise methods. This is by design but indeed causes confusion.

**Why This Happens**: The query builder chain returns a "thenable" object (has a `.then()` method) but isn't a standard Promise. The query only executes when:
- You call `.then()` explicitly
- You use `await` (which internally calls `.then()`)
- You call `.catch()`, `.finally()`, or other promise methods
- You pass it to `Promise.all()` or similar

**The Parallel Execution Issue**: Your example perfectly illustrates the problem. Without `.then()` at the end of each query, they execute sequentially when awaited, not in parallel as many would expect.

**Comparison with Other Libraries**: 
- **Kysely** takes a different approach - it throws an error if you try to await a query builder directly, forcing you to call `.execute()` explicitly. This prevents confusion but requires more verbose code.
- **TypeORM** and **Prisma** have explicit execution methods that make the execution point clearer.
- **Knex** chose the thenable pattern for convenience, allowing `await knex('table').select()` to work directly, but this creates the ambiguity you've highlighted.

**Documentation Improvements Needed**:
1. A prominent section explaining that query builders are "lazy" and don't execute until prompted
2. Clear examples showing the difference between sequential and parallel execution patterns
3. Explicit mention that storing a query in a variable doesn't execute it
4. Warning about reusing query builder instances (as shown in issue #4760 where multiple awaits execute the query multiple times)

**Recommended Patterns**:
```javascript
// For parallel execution - add .then() to trigger immediately
const [friends, requestsToYou, requestsFromYou] = await Promise.all([
  knex('login_users').select('email')
    .innerJoin('users_friends', 'users_friends.user2', 'login_users.id')
    .where({ user1: user, accepted: true })
    .then(), // Triggers execution immediately
  knex('login_users').select('email')
    .innerJoin('users_friends', 'users_friends.user1', 'login_users.id')
    .where({ user2: user, accepted: false })
    .then(),
  knex('login_users').select('email')
    .innerJoin('users_friends', 'users_friends.user2', 'login_users.id')
    .where({ user1: user, accepted: false })
    .then()
]);

// Or use Promise.all without .then() - it will trigger all queries
const [friends, requestsToYou, requestsFromYou] = await Promise.all([
  query1, query2, query3  // Promise.all calls .then() on each
]);
```

As mentioned above, this needs better documentation. A dedicated section on "Understanding Query Execution" would help prevent these misunderstandings. Pull requests improving the documentation in this area would be very welcome.

**Related issues:**
- #1656 - Documentation: async, promise returning nature of migration and seed APIs
- #4760 - Multiple call to .then() behavior documentation  
- #3170 - Replacing .then() blocks with async/await
- #5393 - Knex.raw() throws an error when returned to a promise

**Comment by mercmobily** [comment]
I just closed #5393 (it was a support question) but it gave me the hint that  we should specifically document the dangerous "uncaught exception" scenario that occurs when resolving a Promise with an unexecuted Knex query.

When adding documentation about Knex's thenable/promise behavior, please include prominent warnings about:

1. **Never resolve a Promise directly with an unexecuted query**:
```javascript
// DANGEROUS - causes uncaught exception if query fails
await new Promise(resolve => {
    setTimeout(() => resolve(knex.raw('INVALID SQL')), 100);
});

// SAFE - execute before resolving
await new Promise(async (resolve, reject) => {
    setTimeout(async () => {
        try {
            const result = await knex.raw('SELECT 1');
            resolve(result);
        } catch (err) {
            reject(err);
        }
    }, 100);
});
```

This happens because JavaScript's Promise machinery automatically calls `.then()` on thenables to unwrap them. When the query fails during this automatic execution, the error occurs in Promise internals where it cannot be caught by user code.

This specific pitfall deserves prominent placement in:
- The raw query documentation (`/docs/src/guide/raw.md`)
- The interfaces/promises documentation (`/docs/src/guide/interfaces.md`)
- Any new section about query execution timing

**Comment by mercmobily** [comment]
Closing this in favour of #6252 which includes it
